pkg/convert/convert.go

```diff
@@ -81,6 +81,25 @@ func NewObjectEncoder(os ObjectSchema) (Encoder, error) {
 }
 
 type Decoder interface {
+	// toPropertyValue applies the conversion of the Decoder.
+	//
+	// toPropertyValue should not be called directly. Instead, it should be called
+	// through [decode].
+	//
+	// You *should* do this:
+	//
+	//	var myDecoder Decoder = ...
+	//	var myValue tftypes.Value = ...
+	//	return decode(myDecoder, value)
+	//
+	// You *should not* do this:
+	//
+	//	var myDecoder Decoder = ...
+	//	var myValue tftypes.Value = ...
+	//	return myDecoder.toPropertyValue(value)
+	//
+	// This is because toPropertyValue is not required to handle [tftypes.Value]
+	// modifiers, like secrets and unknown values.
```

Comment:
The Haskell way to emphasize this invariant is some type alias e.g. `type CleanValue tftypes.Value` that only supposed to represent non-secret/unknown cases.

Reply:
I thought of doing something like that. I've been bitten in the past several times where we pass a `type CleanType module.Type` that has stronger guarantees then the underlying type, but the producer of `CleanType` didn't do proper validation (for example, all of Pulumi's `token.*` types). 

I'd want to call it `UnsafeCleanValue` or create an inner package that could enforce validation, and that seemed like to much safety for idiomatic Go.